drivers/uacpi/BindsIo.cpp

```diff
@@ -0,0 +1,226 @@
+
+        return UACPI_STATUS_OK;
+    }
+
+    uacpi_status uacpi_kernel_raw_io_read(uacpi_io_addr address, uacpi_u8 byte_width, uacpi_u64* out_value)
+    {
+#ifdef __x86_64__
+        VALIDATE_(address <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(byte_width <= 4 && byte_width != 3, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(out_value != nullptr, UACPI_STATUS_INVALID_ARGUMENT);
+
+        *out_value = PortRead(address, byte_width);
+        return UACPI_STATUS_OK;
+#else
+        return UACPI_STATUS_UNIMPLEMENTED;
+#endif
+    }
+
+    uacpi_status uacpi_kernel_raw_io_write(uacpi_io_addr address, uacpi_u8 byte_width, uacpi_u64 in_value)
+    {
+#ifdef __x86_64__
+        VALIDATE_(address <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(byte_width <= 4 && byte_width != 3, UACPI_STATUS_INVALID_ARGUMENT);
+
+        PortWrite(address, byte_width, in_value);
+        return UACPI_STATUS_OK;
+#else
+        return UACPI_STATUS_UNIMPLEMENTED;
+#endif
+    }
+
+    uacpi_status uacpi_kernel_pci_read(uacpi_pci_address* address, uacpi_size offset, uacpi_u8 byte_width, uacpi_u64* value)
+    {
+        (void)address; (void)offset; (void)byte_width; (void)value;
+
+        Log("Attempted raw PCI read", LogLevel::Warning);
+        return UACPI_STATUS_UNIMPLEMENTED;
+    }
+
+    uacpi_status uacpi_kernel_pci_write(uacpi_pci_address* address, uacpi_size offset, uacpi_u8 byte_width, uacpi_u64 value)
+    {
+        (void)address; (void)offset; (void)byte_width; (void)value;
+
+        Log("Attempted raw PCI read", LogLevel::Warning);
+        return UACPI_STATUS_UNIMPLEMENTED;
+    }
+
+    union IoMapping //:meme:
+    {
+        struct
+        {
+            uint16_t base;
+            uint16_t length;
+        };
+        uintptr_t packed;
+    };
+
+    uacpi_status uacpi_kernel_io_map(uacpi_io_addr base, uacpi_size len, uacpi_handle* out_handle)
+    {
+#ifdef __x86_64__
+        VALIDATE_(base <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(len <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+
+        IoMapping mapping { .base = (uint16_t)base, .length = (uint16_t)len };
+        *out_handle = reinterpret_cast<uacpi_handle>(mapping.packed);
+
+        return UACPI_STATUS_OK;
+#else
+        return UACPI_STATUS_UNIMPLEMENTED;
+#endif
+    }
+
+    void uacpi_kernel_io_unmap(uacpi_handle handle)
+    { (void)handle; } //no-op
+
+    uacpi_status uacpi_kernel_io_read(uacpi_handle handle, uacpi_size offset, uacpi_u8 byte_width, uacpi_u64* value)
+    {
+#ifdef __x86_64__
+        VALIDATE_(handle != nullptr, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(offset <= 0xFFFF, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(byte_width <= 4 && byte_width != 3, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(value != nullptr, UACPI_STATUS_INVALID_ARGUMENT);
+
+        const IoMapping mapping { .packed = reinterpret_cast<uintptr_t>(handle) };
+        const uint16_t port = mapping.base + offset;
```

Comment:
i see that offset is of type: `uacpi_size` is it guaranteed to be compatible with uint16_t? Are there cases where it could overflow? Or is safe?

Reply:
Hmm in theory there's no guarentee it wouldn't overflow. My assumption is that this is for x86 port io, where the addresses are 16bits, so if the library passes an offset that overflows, it's a bug. Wiuldnt hurt to add an extra validate statement though to be sure.